### ISSUE TITLE
🐞 scrolling becomes delayed system wide after loop is running for a long time

### ISSUE THREAD

**Issue by 8pxl** [MrKai77/Loop#626]
### Bug Description

after loop runs for awhile, the scrolling system wide becomes delayed.

### Steps To Reproduce

The issue arises as long as loop is running (even if it isn't being used)

restarting loop temporarily resolves the issue, but it returns after some time period (usually around 10 minutes, but haven't timed precisely)

### Expected Behavior

Scrolling should not be delayed

### Actual Behavior

Scrolling is delayed after running the app

### macOS Version

Sonama 14.5

### Loop Version

1.1.1

### Final Checks

- [X] My issue is written in English
- [X] My issue title is descriptive
- [X] This is a single bug (multiple bugs should be reported individually)
- [ ] I can help with fixing or developing this issue (tick if you can help).

**Comment by SenpaiHunters** [comment]
What type of scrolling (app(s) other items)? You noted "system wide," have you also made sure its Loop, and not another process, or loop subprocess? And, after 10 mins, isn't a "long time" has Loop ever risen in memory? As, slow scrolling is mostly your CPU or Memory being pegged.

**Comment by 8pxl** [comment]
I waited until the issue occurred again, which happened after about 20 hours (my 10 minute estimate was heavily recency biased...)
I'm not sure how to confirm that Loop is causing this, but I assumed it was since closing Loop resolves the issue.

heres a video of the scrolling delay (i'm just scrolling normally but the page is jumpy)

https://github.com/user-attachments/assets/5ae216ca-05b2-4317-95a3-4ef19648bb3c

screenshot of cpu usage while the issue is occurring
<img width="441" alt="image" src="https://github.com/user-attachments/assets/decc6b97-a8ae-48bd-867b-d374e3fff30b">

and immediately after closing Loop (resolving the issue)

<img width="559" alt="image" src="https://github.com/user-attachments/assets/e4dc4275-16e7-476b-9601-1d78c4c6936b">

**Comment by SenpaiHunters** [comment]
I more wish for you to focus on memory, you'll see Arc use a tonne of it, but normally it'd be CPU/memory or both, in the screenshots shared, it doesn't look like Loop is the cause. WindowServer being the biggest macOS trash in years, finder, mds, spotify, arc, only thing may be `AXVisuals`, but that's very unlikely. The video, it's not really massively clear, and could be just a spike, uhh, we'll have to investigate it more. But, if you can share memory + cpu screenshots when you're having the issues, that'd be awesome.

I'll also add to this, please update to the dev builds, i wish to test if it's also an issue on a newer build, as you've noted you're using `1.1.1` 

![003190-12@2x](https://github.com/user-attachments/assets/272f992c-dfd7-4eee-ace4-69e8c5d0fee2)

**Comment by lukasesch** [comment]
I got a similiar issue, but it's mostly appearing when moving windows around with the mouse, or resizing them.
They are dragging behind the cursor with almost a 1 second delay. 
The longer loop is running, the bigger the issue gets. In my case it needs to be longer than 10 minutes though. Can't put an exact time on it. 
I'll try to take screenshots of memory + cpu when the issue occurs again.

Setup is M1 Macbook Pro (Base Version) attached via USB C Dock to a 4k and 2.5k screen.
Maybe having two different scaled displays could be part of the issue? 

My fix is the same as 8pxl's. Restarting (or closing in general) Loop instantly fixes the issue for some time. 
Dev version is installed since quite some time and the issue was always there.
Loving Loop so I don't mind restarting, nonetheless it starts becoming more annoying.

Thanks if the bug can be found and fixed!

**Comment by greg1075** [comment]
Same here. I experienced the same issue where system animations (e.g. moving / resizing windows, scrolling, etc.) became progressively choppier over time. Rebooting my system would temporarily resolve the problem, but it would gradually reappear the next day (I rarely shut down my MacBook Pro overnight).

Initially, I thought it was a macOS bug. However, I just realized that I haven’t encountered the issue in the past week, which coincides with my switch to Swish. For context, I occasionally alternate between Loop, Swish, and Rectangle Pro...

I run macOS 15.2 (24C101) on an M4 Pro MBP, but had the same problem on the M1 Pro I had before that.

**Comment by greg1075** [comment]
I wanted to confirm the above. Running Loop regularly again, and after a few hours, system starts getting choppy and draggy. Cursor movement, web page scrolling, moving/resizing windows... Quitting Loop fixes the problem instantly.  The problem gets worse over time. Maybe a memory leak of some sort?

**Comment by MrKai77** [comment]
@greg1075 it could indeed be a memory leak. Are you also running the latest version of Loop? We recently fixed a memory leak related to the event monitor, which might address the issue. Additionally, I’ve been exploring some private APIs that avoid mouse event tracking entirely, and I’m considering integrating those into Loop in the future to improve performance further (although I am still seeing it if has any disadvantages compared to our current method).

**Comment by greg1075** [comment]
> [@greg1075](https://github.com/greg1075) it could indeed be a memory leak. Are you also running the latest version of Loop? We recently fixed a memory leak related to the event monitor, which might address the issue. Additionally, I’ve been exploring some private APIs that avoid mouse event tracking entirely, and I’m considering integrating those into Loop in the future to improve performance further (although I am still seeing it if has any disadvantages compared to our current method).

I am running the latest beta version 1.3.0 (1445).

**Comment by SenpaiHunters** [comment]
I'm not going to lie, when i say, i've got 0 idea as to why this is happening for 3 people, and how loop is causing this, and the activity monitor from 8pxl makes no sense because loop is calling none of those programs or events. Plus, it'd be RAM (Memory) i'm interested in, not CPU

**Comment by greg1075** [comment]
> I'm not going to lie, when i say, i've got 0 idea as to why this is happening for 3 people, and how loop is causing this, and the activity monitor from 8pxl makes no sense because loop is calling none of those programs or events. Plus, it'd be RAM (Memory) i'm interested in, not CPU

That is bizarre. It happened to me on two separate systems including a brand new computer i.e. fresh OS , so that pretty much rules out a system-specific issue. It could very well be that more people are running into the issue but haven't reported it? I'm gonna let Loop run and see if I can identify any RAM pressure in the activity monitor.

**Comment by MrKai77** [comment]
> It could very well be that more people are running into the issue but haven't reported it?

I would agree that this is very probable, despite not being able to reproduce it myself.

> I'm gonna let Loop run and see if I can identify any RAM pressure in the activity monitor.

That would be greatly appreciated!

I also suspect that this bug might be related to Loop’s cursor event tracking for window snapping, even though it doesn’t seem directly connected to scrolling. Here’s a build with all of that functionality removed. The option will still appear in Loop’s settings, but it won’t actually do anything: [Loop.zip](https://github.com/user-attachments/files/18547890/Loop.zip)
It would be great if you could test this build to see if the bug still occurs! If it does, I’ll be able to rule out one potential cause of the issue.

**Comment by greg1075** [comment]
The problem started reappearing after about 36 hours. Quitting Loop fixed it immediately. I'm including screenshots of my entire activity monitor with Loop still running (10 total) and after quit (9).

Loop running:

<img width="1379" alt="Image" src="https://github.com/user-attachments/assets/2cc4847d-cd04-48c7-8df3-ca7a70a2e067" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/fa8146b6-a47f-4f34-9ac6-7a57311dd293" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/f3f27389-70d2-4403-9a2b-c55fa6309af8" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/54a6f615-d325-44af-b418-af758d31eed9" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/f7f64f38-b0a5-4acf-9ce4-ebd5860580f0" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/94baec48-5323-4d8c-b13d-f7d54f5b80ea" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/cc1f11d7-1a5e-48ff-b938-b80243be0987" />
<img width="1379" alt="Image" src="https://github.com/user-attachments/assets/623fbd53-1874-44c0-8ff9-81d05551f75c" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/c829ade4-c4c0-4995-8ef1-4c9b5f85a0a2" />

After quit:

<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/51557580-b1a1-4a15-9e0f-6cdbc4d3a280" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/58f45e70-691a-4841-8bd8-4e14fdcc1f82" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/21edb825-3b13-45ec-bc71-44fd758dd4ea" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/54c73a1e-f41b-4b12-9ef4-9bb0d8494287" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/7b15172f-aee2-44a2-a8a4-6867039e00a0" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/898cb187-1735-4825-ac98-0745d938b2b7" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/8fab99b9-a6f1-4a91-8e49-d26ee258b55f" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/71c64851-bd51-4e85-b468-52e4799afb89" />
<img width="1335" alt="Image" src="https://github.com/user-attachments/assets/b9a9bfd4-bceb-4328-822a-31731a84084c" />

**Comment by SenpaiHunters** [comment]
Okay, @greg1075, i can (after i was finished recovering from being blinded...) see Loop shot up to 200 ish mbs of ram, defs should not happen, so there is a leak somewhere, @MrKai77 linked a download above for a stripped version of one of the monitors, can you please download and use that so we can rule out any issues / problems. 

I'm shocked because my loop runtime is over a week and its only at 80mb with all sorts of movements i've done, so maybe theres something macOS version related too, causing something to leak which didnt before.

**Comment by greg1075** [comment]
@SenpaiHunters - Sorry for the onslaught of screenshots. I wasn’t sure exactly what you’d be looking for, so I sent everything.

Also, sorry for not being clear earlier - the activity monitor data above was captured while using the version @MrKai77 linked to above: 1.2.0 (1445).

**Comment by SenpaiHunters** [comment]
Ah, okay, so it's something unrelated to mouse events, are you using any custom keybinds? All i can think is something is not being dislocated or misallocated elsewhere, and have you noticed anything strange while using loop? Currently, the main dev build chain is on `Version 🧪 1.3.0 (1445)`, so even after those changes its something else, interesting (the version will be out because of how builds are done, but don't worry, it'll still work).

Alright, it's something we'll need to debug, thanks for the info and screenshots.

**Comment by greg1075** [comment]
@SenpaiHunters I use 3 custom keybinds, one for window centering which I changed to CMD-[`], Almost Maximize which I changed to CMD-2. and a custom cycle with several (centered) window sizes set to CMD-1.

![Image](https://github.com/user-attachments/assets/58ffeb25-e370-4ac1-8035-24187bf263fc)

**Comment by SenpaiHunters** [comment]
Hey, @lukasesch, @greg1075, @8pxl please refer to my comment in https://github.com/MrKai77/Loop/issues/682#issuecomment-2692176700 for a version you can test to see if it fixes the issue.

**Comment by greg1075** [comment]
> Hey, [@lukasesch](https://github.com/lukasesch), [@greg1075](https://github.com/greg1075), [@8pxl](https://github.com/8pxl) please refer to my comment in [#682 (comment)](https://github.com/MrKai77/Loop/issues/682#issuecomment-2692176700) for a version you can test to see if it fixes the issue.

Downloaded. Will report after a few days of use.

**Comment by greg1075** [comment]
@SenpaiHunters - So far, so good. However, please note that I’ve been traveling and haven’t been using the app as much as I do when I’m at my desk all day. That said, if the memory leak occurs over time rather than after many loops, then after being on for 48 hours, I’m seeing no issue. I’ll report back if anything changes.

**Comment by zenangst** [comment]
The bug happens after you've used Loop (actually invoked commands with it) over time. Fingers crossed that this fixes smashing the bug for good. 🤞

**Comment by greg1075** [comment]
@SenpaiHunters I downloaded the last developmental version that includes the new window animations. Does that version also include the potential fix for the memory leak issue?

**Comment by SenpaiHunters** [comment]
Yes it does, @greg1075, both the animation speed changes and the memory fix are included in the developmental branch, Kai merged those earlier this morning.
